### PR TITLE
fix(CDAP-19974&CDAP-19965): When we retrieve the latest version, assign creation time along with latest="false" for pre 6.8 pipeline

### PR DESCRIPTION
**Pull request by de-lan**
What:

1. Assign creation time along with latest="false" for pre 6.8 "latest" pipeline, this could help fix the history versions UI bug that null created version always shows on top
2. In version history pagination, we should get the actual app instead of implicitly pointing "-SNAPSHOT" to latest. Otherwise the pagination is totally broken

example: null value creationTime version always on top of list
image

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java
Comment (jingtaowu3) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: I feel this method should be named as getApplication.  the original one can be changed to getLatestApplication, or might be even better to refactor it later as getApplication(ApplicationReference appReference)

Comment (sanjanasandeep) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: agreed

Comment (de-lan) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: hmmm, the original one actually gets the application matching passed in ID, only with the exception that it maps `-SNAPSHOT` to latest... I'm out of idea how to name them

Comment (jingtaowu3) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: I see, yeah this is tricky, I think Albert mentioned store shouldn't be the place to handle the "-SNAPSHOT" to latest version. This should be done from Handler. Basically call getApplication(Appid) if version in appId is actual version, else call getApplication(AppReference) which retrieve the latest, something like that

Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: agree with jingtao, should change the `getApplication(ApplicationId)` in this class to actually get that application. There should be another method (`getApplication(ApplicationReference)` or `getLatestApplication(ApplicationReference)`, whatever is more consistent with the rest of the code) to get the latest app, which takes an ApplicationReference as an argument.

This is one of the reasons that special mapping from `-SNAPSHOT` to latest should only happen at the very beginning. Putting it anywhere else makes the logic very difficult to follow, and very easy to introduce bugs.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: should use the creation time from appMeta instead of the current time. Otherwise it's actually possible this time is later than the new version. It also makes it deterministic and easy to unit test, as you know exactly what the creation time should be based on the input.

**Review by albertshau — COMMENTED** [review]
can you add a unit test? There should be something testing the scan call

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java
Comment (de-lan) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: cool, working on that, since we have `ApplicationReference` merged

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java
Comment (de-lan) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: Nice catch, updating this

**Comment by de-lan** [comment]
> can you add a unit test? There should be something testing the scan call

@albertshau hmmm, tried to add the test and found that... unfortunately with the new code, each deployment is adding the creation time in the table, no way to unit test the case that an old `-SNAPSHOT` with null field of creation time...

**Comment by albertshau** [comment]
there's a writeApplicationSerialized() method that directly writes to the underlying table. If we change it to package private and annotate it as `@VisibleForTesting`, we should be able to call it directly from the unit test to write a row with null creation time

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java]: we shouldn't need to be doing this here, the app version is supposed to be accurate. Is that not the case?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: this shouldn't happen here. The only place we should be mapping snapshot to latest is at the very start, in the handler. In those cases, ApplicationReference and ProgramReference should be used. Wherever ApplicationId is used, it should be an actual id.

Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: similar comments everywhere else. I'm not really understanding why this mapping is happening all over the code, is there something that is forcing us to do this?

Comment (de-lan) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: I updated it in the callers of getApplication. OK I'll back traverse more and change it from handler side. And it needs more careful examination I guess.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java
Comment (de-lan) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java]: You're right. This is passing in accurate version. Change reverted

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java
Comment (de-lan) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: This one is reverted, it's already getting latest version from handler

**Comment by de-lan** [comment]
> there's a writeApplicationSerialized() method that directly writes to the underlying table. If we change it to package private and annotate it as `@VisibleForTesting`, we should be able to call it directly from the unit test to write a row with null creation time

Tests added

**Comment by de-lan** [comment]
@albertshau I've added a TODO task to focus on left refactoring work in https://cdap.atlassian.net/browse/CDAP-20033

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/profile/ProfileMetadataMessageProcessor.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/profile/ProfileMetadataMessageProcessor.java]: This class is processing messages about events that happened in the past on a particular entity (some profile got assigned to a specific version of an application). So in this case, there should be no mapping, should use the appId as-is.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberServiceTest.java
Comment (albertshau) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberServiceTest.java]: Shouldn't appDetail.getAppVersion() from a few lines above be the actual correct version?

Something is not behaving correctly if getLatest() is needed here.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java
Comment (albertshau) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java]: nit: this is too long, let's rename to testListRunsWithLegacyRows()

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: it's not good to special case 0 like this just for the unit test. Change with time 0 won't actually be any real row that could ever have been written. Instead, the unit test should be writing a row just like the one before, where all the change information is null.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java
Comment (albertshau) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java]: we can't use this method because it takes a ChangeDetail where the creation time is never null (which is good). Instead, we should be calling the lower level methods in the DefaultStore that directly write the row, passing in null for the creation time.

This test should be moved to DefaultStoreTest instead of in the handler, then you can call package private methods on DefaultStore directly. In general, unit tests should live as close to the relevant logic as possible, testing very focused parts. An edge case like this in the store shouldn't need a test in the handler level, which needs to go through http and the service before getting to the important logic in the store.

**Review by albertshau — COMMENTED** [review]
a couple comments about tests

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java
Comment (de-lan) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java]: hmmm, so it's kinda tricky that because `creationTime` is indexed, in nosql we cannot use `DefaultStore.addApplication` or `AppMetadataStore.writeApplication` to insert a row with null ChangeDetail. I think we have to add the test inside `AppMetadataStoreTest` instead and expose `getApplicationSpecificationTable()` for testing: this way we could insert the wanted row

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/profile/ProfileMetadataMessageProcessor.java
Comment (de-lan) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/profile/ProfileMetadataMessageProcessor.java]: nice catch, will revert

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberServiceTest.java
Comment (de-lan) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberServiceTest.java]: nice catch, will revert

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java
Comment (de-lan) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java]: I'll keep this one for version history and add a new one testing the upgrade pipeline specifically

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java
Comment (de-lan) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java]: @albertshau PTAL now, I've addressed the comments

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java
Comment (de-lan) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/services/http/handlers/AppLifecycleHttpHandlerTest.java]: As offline discussed, will add the test in DefaultStoreTest, adding a null checking when writing the application

**Review by albertshau — APPROVED** [review]
lgtm, assuming CDAP-20033 is addressed in a follow up

**Comment by de-lan** [comment]
> lgtm, assuming CDAP-20033 is addressed in a follow up

Yes it's here: https://github.com/cdapio/cdap/pull/14720
I'll rebase to develop once this one is merged

**Review by sanjanasandeep — APPROVED** [review]

**Pull request merged**
